File: ts-nodejs/src/core/skills.ts
// Read-only Kimi Code skill scanner, a 1:1 port of rust/src/skills.rs (SPEC 21.2).
// No writes, no watchers, no polling: the caller scans once and caches.

import { closeSync, openSync, readSync, readdirSync, statSync } from "node:fs";
import { join } from "node:path";
import { homeDir, kimiHome } from "./credentials.ts";
import {
  rustLines,
  rustStripLeadingBoms,
  rustTrim,
  rustTrimEnd,
  rustTrimMatchesChar,
} from "./json.ts";

export interface SkillInfo {
  id: string;
  name: string;
  description: string;
  /** Group label: "Kimi Code" | "Agents" | "Plugin: <name>" */
  source: string;
}

const FRONTMATTER_BYTES = 4096;

/** `str::split_once(':')` — split at the first colon. */
function splitOnce(input: string, sep: string): [string, string] | null {
  const at = input.indexOf(sep);
  return at < 0 ? null : [input.slice(0, at), input.slice(at + sep.length)];
}

/** name/description out of the YAML frontmatter, read as bytes so a multi-byte
 *  character cut at the 4 KiB boundary degrades to U+FFFD instead of failing. */
export function parseFrontmatterFromBytes(bytes: Uint8Array): [string | null, string | null] {
  const head = new TextDecoder("utf-8", { fatal: false }).decode(bytes.subarray(0, FRONTMATTER_BYTES));
  const lines = rustLines(head);
  const first = lines.shift();
  if (first === undefined) return [null, null];
  if (rustStripLeadingBoms(rustTrim(first)) !== "---") return [null, null];

  let name: string | null = null;
  let description: string | null = null;
  for (const raw of lines) {
    const line = rustTrimEnd(raw);
    if (rustTrim(line) === "---") break;
    const pair = splitOnce(line, ":");
    if (pair === null) continue;
    const value = rustTrimMatchesChar(rustTrimMatchesChar(rustTrim(pair[1]), '"'), "'");
    const key = rustTrim(pair[0]);
    if (key === "name" && name === null) name = value;
    else if (key === "description" && description === null) description = value;
  }
  return [name, description];
}

export function parseFrontmatter(path: string): [string | null, string | null] {
  const bytes = readFileBytes(path);
  if (bytes === null) return [null, null];
  return parseFrontmatterFromBytes(bytes);
}

const isDirectory = (path: string): boolean => {
  try {
    return statSync(path).isDirectory(); // follows symlinks, like Rust's is_dir
  } catch {
    return false;
  }
};

/** `<dir>/<id>/SKILL.md`: the same shape as the Rust `collect`, driven by real
 *  IO because the ported tests use a temp-directory fixture like Rust does. */
export function collectSkills(dir: string, source: string): SkillInfo[] {
  const out: SkillInfo[] = [];
  for (const entry of listDir(dir)) {
    if (!isDirectory(entry.path)) continue;
    const skillMd = join(entry.path, "SKILL.md");
    const bytes = readFileBytes(skillMd);
    if (bytes === null) continue; // no SKILL.md, or unreadable
    const [parsedName, parsedDescription] = parseFrontmatterFromBytes(bytes);
    out.push({
      id: entry.name,
      name: parsedName ?? entry.name, // a missing name falls back to the directory name
      description: parsedDescription ?? "",
      source,
    });
  }
  return out;
}

/** Group label order is by code point ("Agents" < "Kimi Code" < "Plugin: …"),
 *  then a case-insensitive name order — deliberately not localeCompare. */
export function sortSkills(list: SkillInfo[]): SkillInfo[] {
  const compare = (a: string, b: string): number => (a === b ? 0 : a < b ? -1 : 1);
  return list.sort(
    (a, b) =>
      compare(a.source, b.source) || compare(a.name.toLowerCase(), b.name.toLowerCase()),
  );
}

const listDir = (dir: string): { name: string; path: string }[] => {
  try {
    return readdirSync(dir, { withFileTypes: true }).map((entry) => ({
      name: entry.name,
      path: join(dir, entry.name),
    }));
  } catch {
    return [];
  }
};

/** Physically reads at most the first 4 KiB (mirrors Rust's
 *  `File::take(4096)`): a hostile multi-hundred-MB SKILL.md under a managed
 *  plugin directory must not be slurped whole. */
const readFileBytes = (path: string): Uint8Array | null => {
  try {
    const fd = openSync(path, "r");
    try {
      const buf = Buffer.alloc(FRONTMATTER_BYTES);
      const read = readSync(fd, buf, 0, FRONTMATTER_BYTES, 0);
      return new Uint8Array(buf.subarray(0, read));
    } finally {
      closeSync(fd);
    }
  } catch {
    return null;
  }
};

/** The three skill roots, sorted by source then name (SPEC 21.2). */
export function scanSkills(env: NodeJS.ProcessEnv = process.env): SkillInfo[] {
  const out: SkillInfo[] = [];
  const home = homeDir(env);
  if (home === null) return out;
  const kimi = kimiHome(home, env);

  out.push(...collectSkills(join(kimi, "skills"), "Kimi Code"));
  out.push(...collectSkills(join(home, ".agents", "skills"), "Agents"));

  for (const plugin of listDir(join(kimi, "plugins", "managed"))) {
    if (!isDirectory(plugin.path)) continue;
    out.push(...collectSkills(join(plugin.path, "skills"), `Plugin: ${plugin.name}`));
  }
  return sortSkills(out);
}

File: go/internal/core/skills.go
// Read-only Kimi Code skill scanner, a port of rust/src/skills.rs (SPEC 21.2).
// No writes, no watchers, no polling: the caller scans once and caches.
//
// There is no per-skill enabled/disabled state to display: Kimi Code does not
// persist one, and ~/.agents/.skill-lock.json is never read (SPEC 21.2).
package core

import (
	"bytes"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// SkillInfo is one scanned skill.
type SkillInfo struct {
	ID          string
	Name        string
	Description string
	// Source is the group label: "Kimi Code" | "Agents" | "Plugin: <name>"
	Source string
}

const frontmatterBytes = 4096

// ParseFrontmatterFromBytes extracts name/description from the YAML
// frontmatter of a SKILL.md, decoding through LossyDecode so a truncated
// multi-byte character at the 4 KiB cut and stray non-UTF-8 (e.g. GBK) bytes
// degrade to one U+FFFD per maximal ill-formed subpart, exactly like Rust.
// Deliberately a line parser, not a YAML crate: zero new dependencies.
func ParseFrontmatterFromBytes(b []byte) (name, description *string) {
	if len(b) > frontmatterBytes {
		b = b[:frontmatterBytes]
	}
	head := LossyDecode(b)
	lines := RustLines(head)
	if len(lines) == 0 {
		return nil, nil
	}
	if RustStripLeadingBoms(RustTrim(lines[0])) != "---" {
		return nil, nil
	}

	var n, d *string
	for _, raw := range lines[1:] {
		line := RustTrimEnd(raw)
		if RustTrim(line) == "---" {
			break
		}
		key, value, found := strings.Cut(line, ":")
		if !found {
			continue
		}
		// Frontmatter values may be single/double quoted; stripping follows
		// trim_matches (all leading/trailing repeats), not a matching pair.
		v := RustTrimMatchesChar(RustTrimMatchesChar(RustTrim(value), `"`), "'")
		switch RustTrim(key) {
		case "name":
			if n == nil {
				first := v
				n = &first
			}
		case "description":
			if d == nil {
				first := v
				d = &first
			}
		}
	}
	return n, d
}

// ParseFrontmatter reads at most the first 4 KiB of the file — frontmatter
// always sits at the top.
func ParseFrontmatter(path string) (name, description *string) {
	b, ok := readFileBytes(path)
	if !ok {
		return nil, nil
	}
	return ParseFrontmatterFromBytes(b)
}

// CollectSkills reads `<dir>/<id>/SKILL.md` entries, mirroring the Rust
// `collect`. A missing name falls back to the directory name.
func CollectSkills(dir, source string) []SkillInfo {
	var out []SkillInfo
	for _, entry := range listDir(dir) {
		if !isDirectory(entry.path) {
			continue
		}
		b, ok := readFileBytes(filepath.Join(entry.path, "SKILL.md"))
		if !ok {
			continue // no SKILL.md, or unreadable
		}
		parsedName, parsedDescription := ParseFrontmatterFromBytes(b)
		info := SkillInfo{ID: entry.name, Source: source}
		if parsedName != nil {
			info.Name = *parsedName
		} else {
			info.Name = entry.name
		}
		if parsedDescription != nil {
			info.Description = *parsedDescription
		}
		out = append(out, info)
	}
	return out
}

// SortSkills: group label order is by code point ("Agents" < "Kimi Code" <
// "Plugin: …"), then a case-insensitive name order — deliberately not a
// locale collation.
func SortSkills(list []SkillInfo) []SkillInfo {
	sort.SliceStable(list, func(i, j int) bool {
		c := strings.Compare(list[i].Source, list[j].Source)
		if c != 0 {
			return c < 0
		}
		return strings.ToLower(list[i].Name) < strings.ToLower(list[j].Name)
	})
	return list
}

// ScanSkills scans all three skill roots, grouped by source then sorted by
// name (SPEC 21.2).
func ScanSkills(env Env) []SkillInfo {
	out := []SkillInfo{}
	home, ok := HomeDir(env)
	if !ok {
		return out
	}
	kimi := KimiHome(home, env)

	out = append(out, CollectSkills(filepath.Join(kimi, "skills"), "Kimi Code")...)
	out = append(out, CollectSkills(filepath.Join(home, ".agents", "skills"), "Agents")...)

	// Managed plugins: ~/.kimi-code/plugins/managed/<plugin>/skills/<id>/
	for _, plugin := range listDir(filepath.Join(kimi, "plugins", "managed")) {
		if !isDirectory(plugin.path) {
			continue
		}
		out = append(out, CollectSkills(filepath.Join(plugin.path, "skills"), "Plugin: "+plugin.name)...)
	}
	return SortSkills(out)
}

type dirEntry struct {
	name string
	path string
}

func listDir(dir string) []dirEntry {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil
	}
	out := make([]dirEntry, 0, len(entries))
	for _, e := range entries {
		out = append(out, dirEntry{name: e.Name(), path: filepath.Join(dir, e.Name())})
	}
	return out
}

func isDirectory(path string) bool {
	info, err := os.Stat(path) // follows symlinks, like Rust's is_dir
	return err == nil && info.IsDir()
}

// readFileBytes physically reads at most the first 4 KiB (mirrors Rust's
// `File::take(4096)`): a hostile multi-hundred-MB SKILL.md under a managed
// plugin directory must not be slurped whole.
func readFileBytes(path string) ([]byte, bool) {
	f, err := os.Open(path)
	if err != nil {
		return nil, false
	}
	defer f.Close()
	var buf bytes.Buffer
	if _, err := io.CopyN(&buf, f, frontmatterBytes); err != nil && err != io.EOF {
		return nil, false
	}
	return buf.Bytes(), true
}
